Expansion for parallel exits

EthanLuisMcDonough opened the pull request:
This pull request is an extension of #7 that introduces transformation for return statements nested inside parallel constructs.

Richard-Wai commented on gcc/ada/sem_ch6.adb:
```diff
@@ -3475,6 +3482,288 @@ package body Sem_Ch6 is
+           return Boolean
+         is
+            Inside_Outlined : Boolean := True;
+            Scop            : Entity_Id;
```
Nitpicks for the record, I think this variable name could be better.. Maybe something like Test_Scope or Trial_Scope

EthanLuisMcDonough replied:
`Scop` is a variable name used elsewhere in GNAT. See `Build_Constrained_Access_Type`. That said, it might make sense to rename `S` as `Target_Scope`. Do you think that would be clearer?

Richard-Wai replied:
In that case, do as the Romans do!

Richard-Wai commented on gcc/ada/sem_ch6.adb:
```diff
@@ -3475,6 +3482,288 @@ package body Sem_Ch6 is
+               elsif Scop = Scope_Id then
+                  Inside_Outlined := False;
+               end if;
+            end loop;
```
I find this loop somewhat awkward.. I almost expect that

```suggestion
                exit when scope (S) = Scop;
                if Scop = Scope_Id then..
                    ....
            end loop;
            
            return Inside_Outlined;
```
Feels like it should be more appropriate, but otherwise we're saying:
If Scope_ID is NOT in the scope stack then we always return True, regardless if Scope (S) is, but if Scope_ID is in the scope stack, we return False if and only if Scope(S) is above Scope_ID in the stack, and True otherwise.

Anyways I probably haven't spent enough time looking at it, but it's somewhat confusing, and I didn't find the comment super helpful.. Maybe decorate target scope with Scope_ID, and outline function scope with S (if I got that right)

Richard-Wai reviewed (approved):
LGTM

EthanLuisMcDonough commented:
As I've been testing this more, I've realized that `exit when ...` breaks when the condition references a variable defined inside the local scope. We didn't see this issue inside inside the N Queens example because the `when` clause references only variables defined outside the parallel scope. I think I need to move the rewriting logic inside the control flow analysis logic so that the condition gets evaluated in the correct scope.

sttaft commented:
> I think I need to move the rewriting logic inside the control flow analysis logic so that the condition gets evaluated in the correct scope.

Makes sense.  Let's hear it for thorough testing! ;-)

EthanLuisMcDonough commented:
> > I think I need to move the rewriting logic inside the control flow analysis logic so that the condition gets evaluated in the correct scope.
> 
> Makes sense. Let's hear it for thorough testing! ;-)

I think I fixed the issue in #11. Thankfully, I didn't have to move the logic to a different place. I also added some tests for this in  #10.
